File: src/main/java/io/github/malczuuu/iemu/infrastructure/lwm2m/LightControlEnabler.java
package io.github.malczuuu.iemu.infrastructure.lwm2m;

import io.github.malczuuu.iemu.domain.StateDTO;
import io.github.malczuuu.iemu.domain.StateService;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.eclipse.leshan.client.resource.BaseInstanceEnabler;
import org.eclipse.leshan.client.servers.ServerIdentity;
import org.eclipse.leshan.core.model.ObjectModel;
import org.eclipse.leshan.core.node.LwM2mResource;
import org.eclipse.leshan.core.response.ReadResponse;
import org.eclipse.leshan.core.response.WriteResponse;

public class LightControlEnabler extends BaseInstanceEnabler {

  public static LightControlEnabler create(StateService stateService) {
    LightControlEnabler lightControl = new LightControlEnabler(stateService);

    stateService.subscribeOnStateChange(ignored -> lightControl.fireResourcesChange(ON_OFF));
    stateService.subscribeOnDimmerChange(ignored -> lightControl.fireResourcesChange(DIMMER));
    stateService.subscribeOnTimeCounterChange(ignored -> lightControl.fireResourcesChange(ON_TIME));

    return lightControl;
  }

  public static final int OBJECT_ID = 3311;

  private static final int ON_OFF = 5850;
  private static final int DIMMER = 5851;
  private static final int ON_TIME = 5852;
  private static final int CUMULATIVE_ACTIVE_POWER = 5805;
  private static final int POWER_FACTOR = 5820;
  private static final int COLOUR = 5706;
  private static final int SENSOR_UNITS = 5701;
  private static final int APPLICATION_TYPE = 5750;

  private static final Set<Integer> SUPPORTED_RESOURCES =
      new HashSet<>(Arrays.asList(ON_OFF, DIMMER, ON_TIME));

  private final StateService state;

  private LightControlEnabler(StateService state) {
    this.state = state;
  }

  @Override
  public ReadResponse read(ServerIdentity identity, int resourceId) {
    switch (resourceId) {
      case ON_OFF:
        return ReadResponse.success(resourceId, state.getState().getOn());
      case DIMMER:
        return ReadResponse.success(resourceId, state.getState().getDimmer());
      case ON_TIME:
        return ReadResponse.success(resourceId, state.getState().getOnTime());
      case CUMULATIVE_ACTIVE_POWER:
        return ReadResponse.notFound();
    }
    return super.read(identity, resourceId);
  }

  @Override
  public WriteResponse write(ServerIdentity identity, int resourceId, LwM2mResource value) {
    switch (resourceId) {
      case ON_OFF:
        state.changeState(StateDTO.changeOn((Boolean) value.getValue()));
        break;
      case DIMMER:
        state.changeState(StateDTO.changeDimmer(((Long) value.getValue()).intValue()));
        break;
      case ON_TIME:
        state.changeState(StateDTO.changeOnTime(((Long) value.getValue()).intValue()));
        break;
      default:
        return super.write(identity, resourceId, value);
    }
    return WriteResponse.success();
  }

  @Override
  public List<Integer> getAvailableResourceIds(ObjectModel model) {
    return new ArrayList<>(SUPPORTED_RESOURCES);
  }
}

File: src/main/kotlin/io/github/malczuuu/iemu/infra/lwm2m/LightControlEnabler.kt
package io.github.malczuuu.iemu.infra.lwm2m

import io.github.malczuuu.iemu.core.DeviceStateService
import io.github.malczuuu.iemu.core.DeviceStateUpdate
import org.eclipse.leshan.client.resource.BaseInstanceEnabler
import org.eclipse.leshan.client.servers.ServerIdentity
import org.eclipse.leshan.core.model.ObjectModel
import org.eclipse.leshan.core.node.LwM2mResource
import org.eclipse.leshan.core.response.ReadResponse
import org.eclipse.leshan.core.response.WriteResponse

class LightControlEnabler(private val state: DeviceStateService) : BaseInstanceEnabler() {

  init {
    state.onStateChange { fireResourcesChange(ON_OFF) }
    state.onDimmerChange { fireResourcesChange(DIMMER) }
    state.onTimeCounterChange { fireResourcesChange(ON_TIME) }
  }

  override fun read(identity: ServerIdentity, resourceId: Int): ReadResponse =
      when (resourceId) {
        ON_OFF -> ReadResponse.success(resourceId, state.deviceState.on)
        DIMMER -> ReadResponse.success(resourceId, state.deviceState.dimmer.toLong())
        ON_TIME -> ReadResponse.success(resourceId, state.deviceState.onTime)
        CUMULATIVE_ACTIVE_POWER -> ReadResponse.notFound()
        else -> super.read(identity, resourceId)
      }

  override fun write(
      identity: ServerIdentity,
      resourceId: Int,
      value: LwM2mResource,
  ): WriteResponse =
      when (resourceId) {
        ON_OFF -> {
          state.changeDeviceState(DeviceStateUpdate(on = value.value as Boolean))
          WriteResponse.success()
        }

        DIMMER -> {
          state.changeDeviceState(DeviceStateUpdate(dimmer = (value.value as Long).toInt()))
          WriteResponse.success()
        }

        ON_TIME -> {
          state.changeDeviceState(DeviceStateUpdate(onTime = value.value as Long))
          WriteResponse.success()
        }

        else -> super.write(identity, resourceId, value)
      }

  override fun getAvailableResourceIds(model: ObjectModel): List<Int> = SUPPORTED_RESOURCES.toList()

  companion object {
    const val OBJECT_ID: Int = 3311

    private const val ON_OFF = 5850
    private const val DIMMER = 5851
    private const val ON_TIME = 5852
    private const val CUMULATIVE_ACTIVE_POWER = 5805

    private val SUPPORTED_RESOURCES: Set<Int> = setOf(ON_OFF, DIMMER, ON_TIME)

    fun create(deviceStateService: DeviceStateService): LightControlEnabler =
        LightControlEnabler(deviceStateService)
  }
}
